aten/src/ATen/cuda/CUDASparseDescriptors.h

```diff
@@ -0,0 +1,251 @@
+}
+
+} // anonymous namespace
+
+inline cusparseIndexType_t getCuSparseIndexType(const c10::ScalarType& scalar_type) {
+  if (scalar_type == c10::ScalarType::Int) {
+    return CUSPARSE_INDEX_32I;
+  } else if (scalar_type == c10::ScalarType::Long) {
+    return CUSPARSE_INDEX_64I;
+  } else {
+    TORCH_INTERNAL_ASSERT(
+        false,
+        "Cannot convert type ",
+        scalar_type,
+        " to cusparseIndexType.");
+  }
+}
+
+template <typename T, cusparseStatus_t (*destructor)(T*)>
+struct CuSparseDescriptorDeleter {
+  void operator()(T* x) {
+    if (x != nullptr) {
+      TORCH_CUDASPARSE_CHECK(destructor(x));
+    }
+  }
+};
+
+template <typename T, cusparseStatus_t (*destructor)(T*)>
+class CuSparseDescriptor {
+ public:
+  T* descriptor() const {
+    return descriptor_.get();
+  }
+  T* descriptor() {
+    return descriptor_.get();
+  }
+
+ protected:
+  std::unique_ptr<T, CuSparseDescriptorDeleter<T, destructor>> descriptor_;
+};
+
+class TORCH_CUDA_CPP_API CuSparseDnMatDescriptor
+    : public CuSparseDescriptor<cusparseDnMatDescr, &cusparseDestroyDnMat> {
+ public:
+  CuSparseDnMatDescriptor(const Tensor& input) {
+    IntArrayRef input_strides = input.strides();
+    IntArrayRef input_sizes = input.sizes();
+    auto ndim = input.dim();
+    auto rows = input_sizes[ndim - 2];
+    auto cols = input_sizes[ndim - 1];
+
+    bool is_column_major = at::native::is_blas_compatible_column_major_order(input);
+    bool is_row_major = at::native::is_blas_compatible_row_major_order(input);
+    TORCH_INTERNAL_ASSERT(
+        is_column_major || is_row_major,
+        "Expected either row or column major contiguous input.");
+
+    auto leading_dimension = is_row_major ? input_strides[ndim - 2] : input_strides[ndim - 1];
+    auto order = is_row_major ? CUSPARSE_ORDER_ROW : CUSPARSE_ORDER_COL;
+
+    void* values_ptr = input.data_ptr();
+
+    cudaDataType value_type = ScalarTypeToCudaDataType(input.scalar_type());
+    check_supported_cuda_type(value_type);
+
+    cusparseDnMatDescr_t raw_descriptor;
+    TORCH_CUDASPARSE_CHECK(cusparseCreateDnMat(
+        &raw_descriptor,
+        rows,
+        cols,
+        leading_dimension,
+        values_ptr,
+        value_type,
+        order));
+
+    if (ndim > 2) {
+      auto batch_count = at::native::cuda_int_cast(at::native::batchCount(input), "batch_count");
+      auto batch_stride = input_strides[ndim - 3];
+      TORCH_CUDASPARSE_CHECK(cusparseDnMatSetStridedBatch(
+          raw_descriptor, batch_count, batch_stride));
+    }
+
+    descriptor_.reset(raw_descriptor);
+  }
+};
+
+class TORCH_CUDA_CPP_API CuSparseSpMatDescriptor
+    : public CuSparseDescriptor<cusparseSpMatDescr, &cusparseDestroySpMat> {};
+
+class TORCH_CUDA_CPP_API CuSparseSpMatCsrDescriptor
+    : public CuSparseSpMatDescriptor {
+ public:
+  CuSparseSpMatCsrDescriptor(const Tensor& input) {
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.is_sparse_csr());
+    TORCH_INTERNAL_ASSERT_DEBUG_ONLY(input.dim() >= 2);
```

Comment:
As above, let's double check that the user won't run into them when passing the wrong input. You could also write tests for that.

Reply:
It's not possible to create a CSR matrix in PyTorch with `.dim() != 2` currently. There are tests for `torch.sparse_csr_tensor` constructor in `test_sparse_csr.py`.
The `>= 2` condition here to account for the future possible batch extension, I'll remove handling of batched CSR since that code is untested.